### PR TITLE
CoDiPack 2 Update

### PR DESCRIPTION
**Pull request by jblueh**
## Proposed Changes
Update SU2 to a recent version of CoDiPack.

## PR Checklist

- [X] I am submitting my contribution to the develop branch.
- [X] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [X] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [X] I have added a test case that demonstrates my contribution, if necessary.
- [X] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

**Comment by jblueh** [comment]
The changes to SU2 are mostly minor. There are some adaptions regarding external functions, due to interface changes from CoDiPack 1 to CoDiPack 2. The `callPrimalFuncWithADType` is necessary for correct output handling. Passing the external function to it is the intended usage, therefore the encapsulation of the respective code block in a lambda expression. `callPrimalFuncWithADType` also takes care of the tape activity, which was previously spelled out in SU2.

File: SU2_CFD/src/SU2_CFD.cpp [hunk 1]
```diff
@@ -73,6 +73,9 @@ int main(int argc, char *argv[]) {
 #endif
   SU2_MPI::Comm MPICommunicator = SU2_MPI::GetComm();
 
+  /*--- AD initialization ---*/
+  AD::Initialize();
```
[the remark this answers is not in the document]
Comment (jblueh) [reply SU2_CFD/src/SU2_CFD.cpp:77]: Yeah, I think so. The initializations of MPI, OpenMP, and the AD tools should go hand in hand. I have not worked with the python wrapper yet, but my guess would be that it applies there as well.

[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/src/SU2_CFD.cpp:77]: Is there a particular order in which those init functions need to be called? Due to interactions between MPI OpenMP and so on.

[the remark this answers is not in the document]
Comment (jblueh) [reply SU2_CFD/src/SU2_CFD.cpp:77]: The initialization of MeDiPack goes after the initialization of MPI, and using OpenMP requires `MPI_Init_thread`. For everything else, initialize before first use should be sufficient. I have used both initialization order for MPI/MeDiPack and OpenMP/OpDiLib already.

To initialize MPI plus MeDiPack, it is sufficient to call `AMPI_Init` or `AMPI_Init_thread` and allocate what is called `mediTypes` in SU2. SU2's `CMediMPIWrapper::Init` or `CMediMPIWrapper::Init_thread` both achieve this. Internally, MeDiPack calls `AMPI_Init_common` in its init functions. So in C++ SU2, calling `CMediMPIWrapper::Init_AMPI` from the constructor of `CDriver` is redundant with what was already done during MPI initialization.

In python SU2, the MPI initialization via mpi4py does not call the MeDiPack versions of the init functions but the standard ones. Therefore, those need to be complemented by a call to `CMediMPIWrapper::Init_AMPI` that takes care of the missing `AMPI_Init_common` and `mediTypes`.

[the remark this answers is not in the document]
Comment (jblueh) [reply SU2_CFD/src/SU2_CFD.cpp:77]: A call of `omp_initialize` is required if the python wrapper uses a hybrid AD build. It should occur before the first parallel region.

[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/src/SU2_CFD.cpp:77]: Thank you for the explanation :+1: if you want to try moving the init functions to the CDriver ctor that would be great, I can also give it a try during the ongoing work on the wrapper.

**Review by pcarruscag — APPROVED** [review]

File: SU2_CFD/src/SU2_CFD.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jblueh) [reply SU2_CFD/src/SU2_CFD.cpp:77]: Yes, I can take care of the init functions. I'll open a new PR though and merge this one if that's ok?

[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/src/SU2_CFD.cpp:77]: Yep sounds good

**Pull request merged**